ddtrace/profiling/_threading.pyx

```diff
@@ -92,11 +87,17 @@ class _ThreadLink(_thread_link_base):
 
         :param existing_thread_ids: A set of thread ids to keep.
         """
-        with self._lock:
-            # Iterate over a copy of the list of keys since it's mutated during our iteration.
-            for thread_id in list(self._thread_id_to_object.keys()):
-                if thread_id not in existing_thread_ids:
-                    del self._thread_id_to_object[thread_id]
+        # This code clears the thread/object mapping by clearing a copy and swapping it in an atomic operation This is
+        # needed to be able to have this whole class lock-free and avoid concurrency issues.
+        # The fact that is is lock free means we might lose some accuracy, but it's worth the trade-off for speed and simplicity.
+        new_thread_id_to_object_mapping = self._thread_id_to_object.copy()
+        # Iterate over a copy of the list of keys since it's mutated during our iteration.
+        for thread_id in list(new_thread_id_to_object_mapping):
+            if thread_id not in existing_thread_ids:
```

Comment:
does `existing_thread_ids` always have something in them?

if we think it might be often we have cases where it is an empty set, we could always have a short-circuit of:

```python
if not existing_thread_ids:
    self._thread_id_to_object.clear()
```

but maybe not worth it.

Reply:
That's never empty because it's the list of all running threads, and you have at least 2 threads: the main thread, and the stack profiler thread calling this function.

Reply:
👍🏻 makes sense.